
<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Tutorial: Scoring with ANS and other gene signature scoring methods &#8212; ANS  documentation</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=a746c00c" />
    <link rel="stylesheet" type="text/css" href="../_static/sg_gallery.css?v=d2d258e8" />
    <link rel="stylesheet" type="text/css" href="../_static/nbsphinx-code-cells.css?v=2aa19091" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=b3ba4146"></script>
    <script src="../_static/doctools.js?v=888ff710"></script>
    <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'tutorials/basic_scoring';</script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Tutorial: GMM Postprocessing and Hard Labeling" href="gmm_postprocessing_and_hard_labeling.html" />
    <link rel="prev" title="Tutorials" href="../tutorials.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search the docs ..."
         aria-label="Search the docs ..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
<div class="bd-header__inner bd-page-width">
  <label class="sidebar-toggle primary-toggle" for="__primary">
    <span class="fa-solid fa-bars"></span>
  </label>
  
  
  <div class="col-lg-3 navbar-header-items__start">
    
      <div class="navbar-item">

  

<a class="navbar-brand logo" href="../index.html">
  
  
  
  
  
  
    <p class="title logo__title">ANS  documentation</p>
  
</a></div>
    
  </div>
  
  <div class="col-lg-9 navbar-header-items">
    
    <div class="me-auto navbar-header-items__center">
      
        <div class="navbar-item">
<nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../tutorials.html">
                        Tutorials
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../api.html">
                        API
                      </a>
                    </li>
                
  </ul>
</nav></div>
      
    </div>
    
    
    <div class="navbar-header-items__end">
      
        <div class="navbar-item navbar-persistent--container">
          

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
        </div>
      
      
        <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
      
        <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/lciernik/ANS_signature_scoring" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fab fa-github-square fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
      
    </div>
    
  </div>
  
  
    <div class="navbar-persistent--mobile">

 <script>
 document.write(`
   <button class="btn navbar-btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script>
    </div>
  

  
    <label class="sidebar-toggle secondary-toggle" for="__secondary" tabindex="0">
      <span class="fa-solid fa-outdent"></span>
    </label>
  
</div>

    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
      <div class="sidebar-header-items__center">
        
          <div class="navbar-item">
<nav class="navbar-nav">
  <p class="sidebar-header-items__title"
     role="heading"
     aria-level="1"
     aria-label="Site Navigation">
    Site Navigation
  </p>
  <ul class="bd-navbar-elements navbar-nav">
    
                    <li class="nav-item current active">
                      <a class="nav-link nav-internal" href="../tutorials.html">
                        Tutorials
                      </a>
                    </li>
                

                    <li class="nav-item">
                      <a class="nav-link nav-internal" href="../api.html">
                        API
                      </a>
                    </li>
                
  </ul>
</nav></div>
        
      </div>
    
    
    
      <div class="sidebar-header-items__end">
        
          <div class="navbar-item">

<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script></div>
        
          <div class="navbar-item"><ul class="navbar-icon-links navbar-nav"
    aria-label="Icon Links">
        <li class="nav-item">
          
          
          
          
          
          
          
          
          <a href="https://github.com/lciernik/ANS_signature_scoring" title="GitHub" class="nav-link" rel="noopener" target="_blank" data-bs-toggle="tooltip" data-bs-placement="bottom"><span><i class="fab fa-github-square fa-lg" aria-hidden="true"></i></span>
            <span class="sr-only">GitHub</span></a>
        </li>
</ul></div>
        
      </div>
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
<nav class="bd-docs-nav bd-links"
     aria-label="Section Navigation">
  <p class="bd-links__title" role="heading" aria-level="1">Section Navigation</p>
  <div class="bd-toc-item navbar-nav"><ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Tutorial: Scoring with ANS and other gene signature scoring methods</a></li>
<li class="toctree-l1"><a class="reference internal" href="gmm_postprocessing_and_hard_labeling.html">Tutorial: GMM Postprocessing and Hard Labeling</a></li>
<li class="toctree-l1"><a class="reference internal" href="ANS_with_R.html">Tutorial: Scoring with ANS in R</a></li>
</ul>
</div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        
          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item">



<nav aria-label="Breadcrumb">
  <ul class="bd-breadcrumbs">
    
    <li class="breadcrumb-item breadcrumb-home">
      <a href="../index.html" class="nav-link" aria-label="Home">
        <i class="fa-solid fa-home"></i>
      </a>
    </li>
    
    <li class="breadcrumb-item"><a href="../tutorials.html" class="nav-link">Tutorials</a></li>
    
    <li class="breadcrumb-item active" aria-current="page">Tutorial:...</li>
  </ul>
</nav>
</div>
      
    </div>
  
  
</div>
</div>
              
              
              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section id="Tutorial:-Scoring-with-ANS-and-other-gene-signature-scoring-methods">
<h1>Tutorial: Scoring with ANS and other gene signature scoring methods<a class="headerlink" href="#Tutorial:-Scoring-with-ANS-and-other-gene-signature-scoring-methods" title="Permalink to this heading">#</a></h1>
<p>The following package contains the Python implementation of the Adjusted Neighborhood Scoring method, as well as of UCell [1], JASMINE [2] (with likelihood or odds-ratio sub-computation), the scoring approach implemented in the R package <code class="docutils literal notranslate"><span class="pre">Seurat</span></code>[3] proposed Tirosh <em>et al.</em> [4], and the two modification of it <code class="docutils literal notranslate"><span class="pre">Seurat_AG</span></code> and <code class="docutils literal notranslate"><span class="pre">Seurat_LVG</span></code>. We refer to the main article of this work for details on the scoring methods.</p>
<section id="Content-tutorial">
<h2>Content tutorial<a class="headerlink" href="#Content-tutorial" title="Permalink to this heading">#</a></h2>
<p>In this tutorial we show the <strong>usage of each available gene signature scoring methods in this package</strong> and their arguments.</p>
<p>Structure: * Load preprocessed data * Load cell state specific signatures * Score cell state specific signatures * Adjusted Neighborhood Scoring (ANS) * Scanpy Scoring * Seurat Scoring * Seurat_AG Scoring * Seurat_LVG Scoring * JASMINE Scoring * UCell Scoring * Visualizations</p>
<section id="Data">
<h3>Data<a class="headerlink" href="#Data" title="Permalink to this heading">#</a></h3>
<p>We used our preprocessed version of the PBMC dataset and the DGEX list published by Hao et al. 2021 [5]. We will use the preprocessed dataset containing B-cells, monocytes, and natural killer cells. The preprocessed dataset can be additionally downloaded manually <a class="reference external" href="https://drive.google.com/file/d/15DiWGfSoqtt6Fl2tK_0ik-w50rn30LQA/view?usp=drive_link">here</a> and the DGEX list <a class="reference external" href="https://drive.google.com/file/d/1a3Uqky2VZxCxLvGI-soCTUp3lijrfrx7/view?usp=drive_link">here</a>. The raw data can be
downloaded <a class="reference external" href="https://atlas.fredhutch.org/nygc/multimodal-pbmc/">here</a>.</p>
<p><strong>The jupyter notebook automatically downloads the preprocessed dataset and the genes sets, and places them into a ``tut_data`` folder.</strong> Manualld downloaded data should be also placed in the (newly created) <code class="docutils literal notranslate"><span class="pre">tut_data</span></code> foldes.</p>
</section>
<section id="References">
<h3>References<a class="headerlink" href="#References" title="Permalink to this heading">#</a></h3>
<p>[1] Andreatta, Massimo, and Santiago J. Carmona. 2021. “UCell: Robust and Scalable Single-Cell Gene Signature Scoring.” Computational and Structural Biotechnology Journal 19 (June): 3796–98.https://doi.org/10.1016/j.csbj.2021.06.043; <strong>UCell package</strong>: <a class="github reference external" href="https://github.com/carmonalab/UCell">carmonalab/UCell</a></p>
<p>[2] Noureen, Nighat, Zhenqing Ye, Yidong Chen, Xiaojing Wang, and Siyuan Zheng. 2022. “Signature-Scoring Methods Developed for Bulk Samples Are Not Adequate for Cancer Single-Cell RNA Sequencing Data.” eLife 11 (February). <a class="reference external" href="https://doi.org/10.7554/eLife.71994">https://doi.org/10.7554/eLife.71994</a>. <strong>JASMINE R implementation</strong>: <a class="github reference external" href="https://github.com/NNoureen/JASMINE">NNoureen/JASMINE</a></p>
<p>[3] <code class="docutils literal notranslate"><span class="pre">AddModuleScore</span></code> scoring method of R package <code class="docutils literal notranslate"><span class="pre">Seurat</span></code>: <a class="reference external" href="https://satijalab.org/seurat/reference/addmodulescore">https://satijalab.org/seurat/reference/addmodulescore</a></p>
<p>[4] Tirosh, Itay, Benjamin Izar, Sanjay M. Prakadan, Marc H. Wadsworth 2nd, Daniel Treacy, John J. Trombetta, Asaf Rotem, et al. 2016. “Dissecting the Multicellular Ecosystem of Metastatic Melanoma by Single-Cell RNA-Seq.” Science 352 (6282): 189–96. <a class="reference external" href="https://doi.org/10.1126/science.aad0501">https://doi.org/10.1126/science.aad0501</a></p>
<p>[5] Hao, Yuhan, Stephanie Hao, Erica Andersen-Nissen, William M. Mauck 3rd, Shiwei Zheng, Andrew Butler, Maddie J. Lee, et al. 2021. “Integrated Analysis of Multimodal Single-Cell Data.” Cell 184 (13): 3573–87.e29.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">simplefilter</span><span class="p">(</span><span class="n">action</span><span class="o">=</span><span class="s1">&#39;ignore&#39;</span><span class="p">,</span> <span class="n">category</span><span class="o">=</span><span class="ne">FutureWarning</span><span class="p">)</span>

<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="kn">import</span> <span class="nn">gdown</span>

<span class="kn">from</span> <span class="nn">signaturescoring</span> <span class="kn">import</span> <span class="n">score_signature</span>

<span class="kn">from</span> <span class="nn">tut_helper</span> <span class="kn">import</span> <span class="n">get_sigs_from_DGEX_list</span>
<span class="n">sc</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">verbosity</span> <span class="o">=</span> <span class="mi">2</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Create-tutorial-data-folder-and-automatically-download-the-data-if-not-existant.">
<h2>Create tutorial data folder and automatically download the data if not existant.<a class="headerlink" href="#Create-tutorial-data-folder-and-automatically-download-the-data-if-not-existant." title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fn_adata</span> <span class="o">=</span> <span class="s1">&#39;tut_data/pp_pbmc_b_mono_nk.h5ad&#39;</span>
<span class="n">fn_dgex_genes</span> <span class="o">=</span> <span class="s1">&#39;tut_data/DE_by_celltype.csv&#39;</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">exists</span><span class="p">(</span><span class="s1">&#39;./tut_data&#39;</span><span class="p">):</span>
    <span class="n">os</span><span class="o">.</span><span class="n">makedirs</span><span class="p">(</span><span class="s1">&#39;./tut_data&#39;</span><span class="p">)</span>

<span class="c1"># if tutorial data was not downloaded before, it is downloaded</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fn_adata</span><span class="p">):</span>
    <span class="n">URL</span> <span class="o">=</span> <span class="s1">&#39;https://drive.google.com/file/d/15DiWGfSoqtt6Fl2tK_0ik-w50rn30LQA/view?usp=drive_link&#39;</span>
    <span class="n">gdown</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">URL</span><span class="p">,</span> <span class="n">fn_adata</span><span class="p">,</span> <span class="n">fuzzy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

<span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">fn_dgex_genes</span><span class="p">):</span>
    <span class="n">URL</span> <span class="o">=</span> <span class="s1">&#39;https://drive.google.com/file/d/1a3Uqky2VZxCxLvGI-soCTUp3lijrfrx7/view?usp=drive_link&#39;</span>
    <span class="n">gdown</span><span class="o">.</span><span class="n">download</span><span class="p">(</span><span class="n">URL</span><span class="p">,</span> <span class="n">fn_dgex_genes</span><span class="p">,</span> <span class="n">fuzzy</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Load-preprocessed-data">
<h2>Load preprocessed data<a class="headerlink" href="#Load-preprocessed-data" title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="n">fn_adata</span><span class="p">)</span>

<span class="c1">## To avoid errors</span>
<span class="k">if</span> <span class="s1">&#39;log1p&#39;</span> <span class="ow">in</span> <span class="n">adata</span><span class="o">.</span><span class="n">uns_keys</span><span class="p">():</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;log1p&#39;</span><span class="p">][</span><span class="s1">&#39;base&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
<span class="k">else</span><span class="p">:</span>
    <span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;log1p&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;base&#39;</span><span class="p">:</span> <span class="kc">None</span><span class="p">}</span>
</pre></div>
</div>
</div>
<p>The preprocessed dataset contains B-cells, Monocytes and NK-cells.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;celltype.l1&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Mono    43553
NK      14408
B       10613
Name: celltype.l1, dtype: int64
</pre></div></div>
</div>
</section>
<section id="Load-cell-state-specific-signatures">
<h2>Load cell state specific signatures<a class="headerlink" href="#Load-cell-state-specific-signatures" title="Permalink to this heading">#</a></h2>
<p>We create celltype signatures based on the list of published differentially expressed genes per cell type. Because the cell type granularity level is lower in the DGEX genes list, we simply union the DGEX genes of all cell sub-type beloning to our types of interest, i.e., B-cells, Monocytes and NK-cells. The detailed way how to extract the signatures is implemented in the method <code class="docutils literal notranslate"><span class="pre">get_sigs_from_DGEX_list</span></code> of <code class="docutils literal notranslate"><span class="pre">tut_helper.py</span></code>.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">DE_of_celltypes</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">fn_dgex_genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">SG_subtypes</span> <span class="o">=</span> <span class="n">get_sigs_from_DGEX_list</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">DE_of_celltypes</span><span class="p">,</span> <span class="n">remove_overlapping</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Types and their subtypes:
{
    &#34;B&#34;: [
        &#34;B intermediate kappa&#34;,
        &#34;B intermediate lambda&#34;,
        &#34;B memory kappa&#34;,
        &#34;B memory lambda&#34;,
        &#34;B naive kappa&#34;,
        &#34;B naive lambda&#34;,
        &#34;Plasma&#34;,
        &#34;Plasmablast&#34;
    ],
    &#34;Mono&#34;: [
        &#34;CD14 Mono&#34;,
        &#34;CD16 Mono&#34;
    ],
    &#34;NK&#34;: [
        &#34;NK_1&#34;,
        &#34;NK_2&#34;,
        &#34;NK_3&#34;,
        &#34;NK_4&#34;,
        &#34;NK Proliferating&#34;,
        &#34;NK_CD56bright&#34;
    ]
}
WARNING: genes are not in var_names and ignored: [&#39;ABCB9&#39;, &#39;BUB1&#39;, &#39;CAV1&#39;, &#39;CHPF&#39;, &#39;DLGAP5&#39;, &#39;IGF1&#39;, &#39;MYO1D&#39;, &#39;NUGGC&#39;, &#39;PERP&#39;, &#39;UBE2C&#39;]
WARNING: genes are not in var_names and ignored: [&#39;7-Sep&#39;, &#39;CDT1&#39;, &#39;ESCO2&#39;, &#39;GINS2&#39;, &#39;GTSE1&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">k</span><span class="p">,</span><span class="n">v</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Signature for subtype </span><span class="si">{</span><span class="n">k</span><span class="si">}</span><span class="s1"> contains </span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">v</span><span class="p">)</span><span class="si">}</span><span class="s1"> genes.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Signature for subtype B contains 488 genes.
Signature for subtype Mono contains 382 genes.
Signature for subtype NK contains 243 genes.
</pre></div></div>
</div>
</section>
<section id="Score-cell-state-specific-signatures">
<h2>Score cell state specific signatures<a class="headerlink" href="#Score-cell-state-specific-signatures" title="Permalink to this heading">#</a></h2>
<p>Next we show how to score cells with each method and avaialble parameters. We will score all signatures in <code class="docutils literal notranslate"><span class="pre">SG_subtypes</span></code>.</p>
<p>All scoring methods can be called via the method <code class="docutils literal notranslate"><span class="pre">score_signature</span></code> of our package:</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">signaturescoring</span> <span class="k">as</span> <span class="nn">ssc</span>

<span class="n">ssc</span><span class="o">.</span><span class="n">score_signature</span><span class="p">(</span>
    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_signature</span><span class="p">,</span>
    <span class="n">method</span><span class="o">=</span><span class="s1">&#39;[desired scoring method]&#39;</span><span class="p">,</span>
    <span class="n">score_name</span><span class="o">=</span><span class="s1">&#39;scores&#39;</span><span class="p">,</span>
    <span class="n">gene_pool</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">df_mean_var</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
    <span class="n">copy</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
    <span class="o">**</span><span class="n">kwarg</span>                                 <span class="c1"># Scoring method specific keyword arguments</span>
<span class="p">)</span>
</pre></div>
</div>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">adata</span></code></p></td>
<td><p>required arg</p></td>
<td><p>AnnData object containing the log-normalized gene expression.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">gene_list</span></code></p></td>
<td><p>required arg</p></td>
<td><p>A list of genes,i.e., gene expression signature, for which the cells are scored
for.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">method</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">adjusted_neighborhood_scoring</span></code></p></td>
<td><p>Scoring method to use. One of [‘adjusted_neighborhood_scoring’,
‘seurat_scoring’, ‘seurat_ag_scoring’,’seurat_lvg_scoring’, ‘scanpy_scoring’,
‘jasmine_scoring’, ‘ucell_scoring’]</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">score_name</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">scores</span></code></p></td>
<td><p>Column name for scores stored in <code class="docutils literal notranslate"><span class="pre">.obs</span></code> of <code class="docutils literal notranslate"><span class="pre">adata</span></code>.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">gene_pool</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">None</span></code></p></td>
<td><p>The pool of genes out of which control genes can be selected. If it is <code class="docutils literal notranslate"><span class="pre">None</span></code>,
<code class="docutils literal notranslate"><span class="pre">gene_pool</span></code> is set to remaining genes. NOTE: Only used in ANS, Scanpy, Seurat,
Seurat_AG, Seurat_LVG</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">df_mean_var</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">None</span></code></p></td>
<td><p>A pandas DataFrame containing the average expression (and variance) for each
gene in the dataset. If <code class="docutils literal notranslate"><span class="pre">df_mean_var</span></code> is <code class="docutils literal notranslate"><span class="pre">None</span></code>, the average gene expression
and variance is computed during gene signature scoring. NOTE: Only used in ANS,
Seurat, Seurat_AG, Seurat_LVG</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">copy</span></code></p></td>
<td><p>False</p></td>
<td><p>Indicates whether original or a copy of <code class="docutils literal notranslate"><span class="pre">adata</span></code> is modified.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">use_raw</span></code></p></td>
<td><p>False</p></td>
<td><p>Whether to compute gene signature score on raw data stored in <code class="docutils literal notranslate"><span class="pre">.raw</span></code> attribute
of <code class="docutils literal notranslate"><span class="pre">adata</span></code>.</p></td>
</tr>
</tbody>
</table>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1">## Helper array to store all names of the newly created columns</span>
<span class="n">all_score_names</span> <span class="o">=</span> <span class="p">[]</span>
</pre></div>
</div>
</div>
<section id="Adjusted-Neighborhood-Scoring-(ANS)">
<h3>Adjusted Neighborhood Scoring (ANS)<a class="headerlink" href="#Adjusted-Neighborhood-Scoring-(ANS)" title="Permalink to this heading">#</a></h3>
<p>Our proposed method.</p>
<p><em>Method specific arguments:</em></p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ctrl_size</span></code></p></td>
<td><p>100</p></td>
<td><p>The number of control genes selected for each gene in the gene_list.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">remove_genes_with_invalid_control_set</span></code></p></td>
<td><p>True</p></td>
<td><p>If true, the scoring method removes genes from the <code class="docutils literal notranslate"><span class="pre">gene_list</span></code> for which no
optimal control set can be computed, i.e., if a gene belongs to the
<code class="docutils literal notranslate"><span class="pre">ctrl_size/2</span></code> genes with the largest average expression.</p></td>
</tr>
</tbody>
</table>
<p>For more details, see the implementation in <code class="docutils literal notranslate"><span class="pre">signaturescoring/scoring_methods/adjusted_neighborhood_scoring.py</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;ANS_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with ANS</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;adjusted_neighborhood_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">ctrl_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;ANS_B&#39;
computing score &#39;ANS_B&#39;
    finished (0:00:04)
Scoring for gene type and storing scores in &#39;ANS_Mono&#39;
computing score &#39;ANS_Mono&#39;
    finished (0:00:02)
Scoring for gene type and storing scores in &#39;ANS_NK&#39;
computing score &#39;ANS_NK&#39;
    finished (0:00:03)
</pre></div></div>
</div>
</section>
<section id="Scanpy-scoring">
<h3>Scanpy scoring<a class="headerlink" href="#Scanpy-scoring" title="Permalink to this heading">#</a></h3>
<p>Selecting <code class="docutils literal notranslate"><span class="pre">method='scanpy_scoring'</span></code> will call the original <code class="docutils literal notranslate"><span class="pre">score_genes</span></code> method of the Scanpy package. See details on the method <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.tl.score_genes.html">here</a>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Scanpy_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with Scanpy</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;scanpy_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">ctrl_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                    <span class="n">n_bins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;Scanpy_B&#39;
computing score &#39;Scanpy_B&#39;
    finished (0:00:01)
Scoring for gene type and storing scores in &#39;Scanpy_Mono&#39;
computing score &#39;Scanpy_Mono&#39;
    finished (0:00:01)
Scoring for gene type and storing scores in &#39;Scanpy_NK&#39;
computing score &#39;Scanpy_NK&#39;
    finished (0:00:01)
</pre></div></div>
</div>
</section>
<section id="Seurat">
<h3>Seurat<a class="headerlink" href="#Seurat" title="Permalink to this heading">#</a></h3>
<p>The following three chapters show the usage of the scoring method implemented in the R Package <code class="docutils literal notranslate"><span class="pre">`Seurat</span></code> &lt;<a class="reference external" href="https://satijalab.org/seurat/reference/addmodulescore">https://satijalab.org/seurat/reference/addmodulescore</a>&gt;`__ and first described by Tirosh <em>et al.</em> (2016)and the two modifications of it.</p>
<p><em>Method specific arguments:</em></p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ctrl_size</span></code></p></td>
<td><p>100</p></td>
<td><p>The number of control genes selected for each gene in the gene_list.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">n_bins</span></code></p></td>
<td><p>25</p></td>
<td><p>The number of average gene expression bins to use.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">random_state</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">None</span></code></p></td>
<td><p>Seed for random state. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, no global seed is set.</p></td>
</tr>
</tbody>
</table>
<p>For more details, see the implementation in <code class="docutils literal notranslate"><span class="pre">signaturescoring/scoring_methods/seurat_scoring.py</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Seurat_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with Seurat</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;seurat_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">ctrl_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                    <span class="n">n_bins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;Seurat_B&#39;
computing score &#39;Seurat_B&#39;
    finished (0:00:02)
Scoring for gene type and storing scores in &#39;Seurat_Mono&#39;
computing score &#39;Seurat_Mono&#39;
    finished (0:00:01)
Scoring for gene type and storing scores in &#39;Seurat_NK&#39;
computing score &#39;Seurat_NK&#39;
    finished (0:00:01)
</pre></div></div>
</div>
</section>
<section id="Seurat_AG">
<h3>Seurat_AG<a class="headerlink" href="#Seurat_AG" title="Permalink to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Seurat_AG</span></code> is a modification of the <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> method and uses all non-signature genes of an expression bin as control. Thus does <strong>not</strong> require the <code class="docutils literal notranslate"><span class="pre">ctrl_size</span></code> argunment and results in the same control set for all signature genes landing in the same expression bin.</p>
<p><em>Method specific arguments:</em></p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">n_bins</span></code></p></td>
<td><p>25</p></td>
<td><p>The number of average gene expression bins to use.</p></td>
</tr>
</tbody>
</table>
<p>For more details, see the implementation in <code class="docutils literal notranslate"><span class="pre">signaturescoring/scoring_methods/seurat_ag_scoring.py</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Seurat_AG_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with Seurat_AG</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;seurat_ag_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">n_bins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;Seurat_AG_B&#39;
computing score &#39;Seurat_AG_B&#39;
    finished (0:00:04)
Scoring for gene type and storing scores in &#39;Seurat_AG_Mono&#39;
computing score &#39;Seurat_AG_Mono&#39;
    finished (0:00:02)
Scoring for gene type and storing scores in &#39;Seurat_AG_NK&#39;
computing score &#39;Seurat_AG_NK&#39;
    finished (0:00:04)
</pre></div></div>
</div>
</section>
<section id="Seurat_LVG">
<h3>Seurat_LVG<a class="headerlink" href="#Seurat_LVG" title="Permalink to this heading">#</a></h3>
<p><code class="docutils literal notranslate"><span class="pre">Seurat_LVG</span></code> is a modification of the <code class="docutils literal notranslate"><span class="pre">Seurat</span></code> method and uses <code class="docutils literal notranslate"><span class="pre">ctrl_size</span></code> genes per expression bin with smallest variance as control. all non-signature genes of an expression bin as control. Thus resulting in the same control set for all signature genes landing in the same expression bin. The method makes use of the <code class="docutils literal notranslate"><span class="pre">highly_variable_genes</span></code> method provided by Scanpy. Find more details on the Scanpy method
<a class="reference external" href="https://scanpy.readthedocs.io/en/stable/generated/scanpy.pp.highly_variable_genes.html">here</a>.</p>
<p><em>Method specific arguments:</em></p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">ctrl_size</span></code></p></td>
<td><p>100</p></td>
<td><p>The number of control genes selected for each gene in the gene_list.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">n_bins</span></code></p></td>
<td><p>25</p></td>
<td><p>The number of average gene expression bins to use.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">lvg_computation_version</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">v1</span></code></p></td>
<td><p>The version of the least variable genes selection defines if the genes with the
smallest dispersion are chosen directly from an expression bin (v1) or whether
the expressions are binned a second round (v2).</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">lvg_computation_method</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">seurat</span></code></p></td>
<td><p>Indicates which method should be used to compute the least variable genes. We
can use <code class="docutils literal notranslate"><span class="pre">seurat</span></code> or <code class="docutils literal notranslate"><span class="pre">cell_ranger</span></code>. See reference
[here].(<a class="reference external" href="https://scanpy.readthedocs.io/en/stable">https://scanpy.readthedocs.io/en/stable</a>
/generated/scanpy.pp.highly_variable_genes.html#scanpy.pp.highly_variable_genes)</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">nr_norm_bins</span></code></p></td>
<td><p>5</p></td>
<td><p>If <code class="docutils literal notranslate"><span class="pre">lvg_computation_version='v2'</span></code>, we need to define the number of subbins
used.</p></td>
</tr>
</tbody>
</table>
<p>For more details, see the implementation in <code class="docutils literal notranslate"><span class="pre">signaturescoring/scoring_methods/seurat_lvg_scoring.py</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Seurat_LVG_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with Seurat_LVG</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;seurat_lvg_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">ctrl_size</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
                    <span class="n">n_bins</span><span class="o">=</span><span class="mi">25</span><span class="p">,</span>
                    <span class="n">lvg_computation_version</span><span class="o">=</span><span class="s2">&quot;v1&quot;</span><span class="p">,</span>
                    <span class="n">lvg_computation_method</span><span class="o">=</span><span class="s2">&quot;seurat&quot;</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;Seurat_LVG_B&#39;
computing score &#39;Seurat_LVG_B&#39;
extracting highly variable genes
    finished (0:00:08)
    finished (0:00:11)
Scoring for gene type and storing scores in &#39;Seurat_LVG_Mono&#39;
computing score &#39;Seurat_LVG_Mono&#39;
extracting highly variable genes
    finished (0:00:08)
    finished (0:00:11)
Scoring for gene type and storing scores in &#39;Seurat_LVG_NK&#39;
computing score &#39;Seurat_LVG_NK&#39;
extracting highly variable genes
    finished (0:00:08)
    finished (0:00:12)
</pre></div></div>
</div>
</section>
<section id="JASMINE-scoring">
<h3>JASMINE scoring<a class="headerlink" href="#JASMINE-scoring" title="Permalink to this heading">#</a></h3>
<p>We implemented the JASMINE gene signature scoring method proposed by Andreatta <em>et al.</em> 2021 in Python and can be used by our package. The scoring method proposes the computation of the scores via the <code class="docutils literal notranslate"><span class="pre">likelihood</span></code> or the <code class="docutils literal notranslate"><span class="pre">odds-ratio</span></code>. We refer to the original article for detailed distinction of the methods.</p>
<p><em>Methods specific arguments:</em></p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">score_method</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">likelihood</span></code></p></td>
<td><p>The method describes, which submethod of enrichment value computation should be
used: <code class="docutils literal notranslate"><span class="pre">likelihood</span></code> or <code class="docutils literal notranslate"><span class="pre">oddsratio</span></code>.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">bs</span></code></p></td>
<td><p>500</p></td>
<td><p>The number of cells in a processing batch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">random_state</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">None</span></code></p></td>
<td><p>Seed for random state. If <code class="docutils literal notranslate"><span class="pre">None</span></code> not global seed is set.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">joblib_kwargs</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">{'n_jobs':</span> <span class="pre">4}</span></code></p></td>
<td><p>Keyword argument for parallel execution with joblib.</p></td>
</tr>
</tbody>
</table>
<p>For more details, see the implementation in <code class="docutils literal notranslate"><span class="pre">signaturescoring/scoring_methods/jasmine_scoring.py</span></code></p>
<section id="Jasmine_LH">
<h4>Jasmine_LH<a class="headerlink" href="#Jasmine_LH" title="Permalink to this heading">#</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Jasmine_LH_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with Jasmine_LH</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;jasmine_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">score_method</span><span class="o">=</span><span class="s1">&#39;likelihood&#39;</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;Jasmine_LH_B&#39;
computing score &#39;Jasmine_LH_B&#39;
    finished (0:00:16)
Scoring for gene type and storing scores in &#39;Jasmine_LH_Mono&#39;
computing score &#39;Jasmine_LH_Mono&#39;
    finished (0:00:16)
Scoring for gene type and storing scores in &#39;Jasmine_LH_NK&#39;
computing score &#39;Jasmine_LH_NK&#39;
    finished (0:00:14)
</pre></div></div>
</div>
</section>
<section id="Jasmine_OR">
<h4>Jasmine_OR<a class="headerlink" href="#Jasmine_OR" title="Permalink to this heading">#</a></h4>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;Jasmine_OR_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with Jasmine_OR</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;jasmine_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">score_method</span><span class="o">=</span><span class="s1">&#39;oddsratio&#39;</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;Jasmine_OR_B&#39;
computing score &#39;Jasmine_OR_B&#39;
    finished (0:00:15)
Scoring for gene type and storing scores in &#39;Jasmine_OR_Mono&#39;
computing score &#39;Jasmine_OR_Mono&#39;
    finished (0:00:16)
Scoring for gene type and storing scores in &#39;Jasmine_OR_NK&#39;
computing score &#39;Jasmine_OR_NK&#39;
    finished (0:00:15)
</pre></div></div>
</div>
</section>
</section>
<section id="UCell-scoring">
<h3>UCell scoring<a class="headerlink" href="#UCell-scoring" title="Permalink to this heading">#</a></h3>
<p>We implemented the UCell gene signature scoring method proposed by Noureen <em>et al.</em> 2022 in Python and can be used by our package. We refer to the original article for detailed distinction of the methods.</p>
<p><em>Methods specific arguments:</em></p>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Argument</p></th>
<th class="head"><p>Default</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">maxRank</span></code></p></td>
<td><p>1500</p></td>
<td><p>Cutoff for maximum rank allowed.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">bs</span></code></p></td>
<td><p>500</p></td>
<td><p>The number of cells in a processing batch.</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">random_state</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">None</span></code></p></td>
<td><p>Seed for random state. If <code class="docutils literal notranslate"><span class="pre">None</span></code> not global seed is set.</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">joblib_kwargs</span></code></p></td>
<td><p><code class="docutils literal notranslate"><span class="pre">{'n_jobs':</span> <span class="pre">4}</span></code></p></td>
<td><p>Keyword argument for parallel execution with joblib.</p></td>
</tr>
</tbody>
</table>
<p>For more details, see the implementation in <code class="docutils literal notranslate"><span class="pre">signaturescoring/scoring_methods/ucell_scoring.py</span></code></p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="k">for</span> <span class="n">gene_type</span><span class="p">,</span> <span class="n">gene_list</span> <span class="ow">in</span> <span class="n">SG_subtypes</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
    <span class="c1"># defining name of the scores column in .obs</span>
    <span class="n">score_name</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;UCell_</span><span class="si">{</span><span class="n">gene_type</span><span class="si">}</span><span class="s1">&#39;</span>

    <span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Scoring for gene type and storing scores in </span><span class="se">\&#39;</span><span class="si">{</span><span class="n">score_name</span><span class="si">}</span><span class="se">\&#39;</span><span class="s1">&#39;</span><span class="p">)</span>

    <span class="c1"># scoring with UCell</span>
    <span class="n">score_signature</span><span class="p">(</span><span class="n">method</span><span class="o">=</span><span class="s1">&#39;ucell_scoring&#39;</span><span class="p">,</span>
                    <span class="n">adata</span><span class="o">=</span><span class="n">adata</span><span class="p">,</span>
                    <span class="n">gene_list</span><span class="o">=</span><span class="n">gene_list</span><span class="p">,</span>
                    <span class="n">maxRank</span><span class="o">=</span> <span class="mi">1500</span><span class="p">,</span>
                    <span class="n">score_name</span><span class="o">=</span><span class="n">score_name</span>
                    <span class="p">)</span>

    <span class="n">all_score_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">score_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Scoring for gene type and storing scores in &#39;UCell_B&#39;
computing score &#39;UCell_B&#39;
    finished (0:00:20)
Scoring for gene type and storing scores in &#39;UCell_Mono&#39;
computing score &#39;UCell_Mono&#39;
    finished (0:00:20)
Scoring for gene type and storing scores in &#39;UCell_NK&#39;
computing score &#39;UCell_NK&#39;
    finished (0:00:20)
</pre></div></div>
</div>
</section>
</section>
<section id="Visualizing-scores">
<h2>Visualizing scores<a class="headerlink" href="#Visualizing-scores" title="Permalink to this heading">#</a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
computing PCA
    with n_comps=50
    finished (0:00:47)
computing neighbors
    using &#39;X_pca&#39; with n_pcs = 50
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/local/lciernik/miniconda3/envs/test_env/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
    finished (0:00:22)
computing UMAP
    finished (0:00:30)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;celltype.l1&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype.l2&#39;</span><span class="p">,</span> <span class="s1">&#39;celltype.l3&#39;</span><span class="p">]</span><span class="o">+</span><span class="n">all_score_names</span><span class="p">,</span> <span class="n">ncols</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/local/lciernik/miniconda3/envs/test_env/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:391: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/local/lciernik/miniconda3/envs/test_env/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:391: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
/local/lciernik/miniconda3/envs/test_env/lib/python3.8/site-packages/scanpy/plotting/_tools/scatterplots.py:391: UserWarning: No data for colormapping provided via &#39;c&#39;. Parameters &#39;cmap&#39; will be ignored
  cax = scatter(
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/tutorials_basic_scoring_34_1.png" src="../_images/tutorials_basic_scoring_34_1.png" />
</div>
</div>
</section>
<section id="Store-adata-for-following-tutorials">
<h2>Store adata for following tutorials<a class="headerlink" href="#Store-adata-for-following-tutorials" title="Permalink to this heading">#</a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">write_h5ad</span><span class="p">(</span><span class="s1">&#39;tut_data/pp_pbmc_b_mono_nk_scored.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
</section>


                </article>
              
              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../tutorials.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Tutorials</p>
      </div>
    </a>
    <a class="right-next"
       href="gmm_postprocessing_and_hard_labeling.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Tutorial: GMM Postprocessing and Hard Labeling</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">

  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> On this page
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Content-tutorial">Content tutorial</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Data">Data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#References">References</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Create-tutorial-data-folder-and-automatically-download-the-data-if-not-existant.">Create tutorial data folder and automatically download the data if not existant.</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Load-preprocessed-data">Load preprocessed data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Load-cell-state-specific-signatures">Load cell state specific signatures</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Score-cell-state-specific-signatures">Score cell state specific signatures</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Adjusted-Neighborhood-Scoring-(ANS)">Adjusted Neighborhood Scoring (ANS)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Scanpy-scoring">Scanpy scoring</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Seurat">Seurat</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Seurat_AG">Seurat_AG</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#Seurat_LVG">Seurat_LVG</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#JASMINE-scoring">JASMINE scoring</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#Jasmine_LH">Jasmine_LH</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#Jasmine_OR">Jasmine_OR</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#UCell-scoring">UCell scoring</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Visualizing-scores">Visualizing scores</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#Store-adata-for-following-tutorials">Store adata for following tutorials</a></li>
</ul>
  </nav></div>

  <div class="sidebar-secondary-item">

  <div class="tocsection sourcelink">
    <a href="../_sources/tutorials/basic_scoring.ipynb.txt">
      <i class="fa-solid fa-file-lines"></i> Show Source
    </a>
  </div>
</div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
          </footer>
        
      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
<div class="bd-footer__inner bd-page-width">
  
    <div class="footer-items__start">
      
        <div class="footer-item">

  <p class="copyright">
    
      © Copyright 2023, ANS contributors.
      <br/>
    
  </p>
</div>
      
        <div class="footer-item">

  <p class="sphinx-version">
    Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.1.2.
    <br/>
  </p>
</div>
      
    </div>
  
  
  
    <div class="footer-items__end">
      
        <div class="footer-item">
<p class="theme-version">
  Built with the <a href="https://pydata-sphinx-theme.readthedocs.io/en/stable/index.html">PyData Sphinx Theme</a> 0.14.4.
</p></div>
      
    </div>
  
</div>

  </footer>
  </body>
</html>